### PR TITLE
Add handling of alt-scores for subjects, assessments and exams.

### PR DESCRIPTION
**Pull request by malaffoon**
Yes, this is an obnoxiously large PR, sorry. In an attempt to stop it from getting even larger, there is some hack code in place so migrate acts as before (only OVERALL, CLAIM data is migrated). That will be removed/changed in the next chunk of work.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -21,8 +21,8 @@ plugins {
 apply plugin: 'io.spring.dependency-management'
 
 // Set version dependencies. This allows the cli to force version, e.g. `-Pcommon=1.2.0-SNAPSHOT`
-String rdwCommonVersion = project.hasProperty("common") ? project.property("common") : "1.4.0-108"
-String rdwSchemaVersion = project.hasProperty("schema") ? project.property("schema") : "1.4.0-411"
+String rdwCommonVersion = project.hasProperty("common") ? project.property("common") : "1.4.0-SNAPSHOT"
+String rdwSchemaVersion = project.hasProperty("schema") ? project.property("schema") : "1.4.0-SNAPSHOT"
```
Comment (malaffoon) [inline build.gradle]: will set when i have build numbers

File: common/src/main/resources/application-ingest-common.yml [hunk 2]
```diff
@@ -1,10 +1,4 @@
 validation:
-  scaleScore:
```
Comment (malaffoon) [inline common/src/main/resources/application-ingest-common.yml]: The scale score limits are now pulled from subject and assessment configuration.

File: build.gradle (hunk 1 above)
Comment (esotrope) [inline build.gradle]: does this effectively pull latest? is this what we want?

File: exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Assessment.java [hunk 3]
```diff
@@ -2,15 +2,14 @@
...
-import javax.validation.constraints.NotNull;
 import java.util.List;
 import java.util.Map;
 
 import static com.google.common.collect.Lists.newArrayList;
 import static com.google.common.collect.Maps.newHashMap;
 
 /**
- * An assessment definition
+ * An assessment definition used by the exam processor.
```
Comment (malaffoon) [inline exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Assessment.java]: It's a bit confusing because there are similarly named data models in multiple modules. I added this kind of comment to indicate that this is the version needed by the exam processor.

File: exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/ExamScore.java [hunk 4]
```diff
@@ -2,24 +2,28 @@
...
 
 /**
- * An {@link Exam}'s claim scoring data
+ * An {@link Exam}'s alt/claim scoring data.
+ * <p>
+ * Note: the overall score is part of the Exam itself, this class is for
+ * storing additional (and optional) alt-scores and claim scores.
+ * </p>
  */
-public class ExamClaim {
```
Comment (malaffoon) [inline exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/ExamScore.java]: A LOT of the changes are removing claim-specificity in names and methods.

File: exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Assessment.java [hunk 5]
```diff
@@ -19,20 +18,24 @@
...
+    /**
+     * @return subject scores, i.e. code,id pairs for claims and alt-scores
+     */
+    public List<AssessmentScore> getScores() {
+        return scores == null ? newArrayList() : Lists.newArrayList(scores);
     }
 
-    @NotNull
     public Map<String, Integer> getItems() {
         return items == null ? newHashMap() : newHashMap(items);
```
Comment (esotrope) [inline exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Assessment.java]: we have the convention in reporting to use "safeImmuableOf()" wondering if it's worth adding a `safeCopyOf` method for these.

File: exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/ExamScore.java [hunk 6]
```diff
@@ -2,24 +2,28 @@
...
 
     private Long id;
-    private int claimId;
+    private int subjectScoreId;
     private Double scaleScore;
     private Double scaleScoreStdErr;
     private Double thetaScore;
     private Double thetaScoreStdErr;
-    private Integer category;
+    private Integer performanceLevel;
```
Comment (malaffoon) [inline exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/ExamScore.java]: Although SB uses the term "category" for claims, now that this can be a claim or alt-score, i changed it to the more generic performance level.

File: exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/AssessmentRepository.java [hunk 7]
```diff
@@ -10,8 +10,8 @@
     /**
      * Finds an {@link Assessment} by the natural id.
      *
-     * @param id the natural id
+     * @param naturalId the natural id
      * @return the found {@link Assessment}; null if not found
      */
-    Assessment findOneByNaturalId(String id);
+    Assessment findOneByNaturalId(String naturalId);
```
Comment (malaffoon) [inline exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/AssessmentRepository.java]: ok, not every change was strictly necessary.

File: exam-processor/src/main/resources/application.sql.yml [hunk 8]
```diff
@@ -66,17 +66,28 @@ sql:
...
+      findAllForAssessment: >-
+        SELECT 1 as score_type_id, NULL as subject_score_id, 'OVERALL' as code, min_score, max_score
+          FROM asmt_score
+          WHERE subject_score_id IS NULL and asmt_id = :asmt_id
+        UNION ALL
+        SELECT s.score_type_id, s.id as subject_score_id, s.code, a.min_score, a.max_score
+          FROM subject_score s
+            JOIN asmt ON s.subject_id = asmt.subject_id AND s.asmt_type_id = asmt.type_id
+            LEFT JOIN asmt_score a ON s.id = a.subject_score_id AND asmt.id = a.asmt_id
+          WHERE asmt.id = :asmt_id
```
Comment (malaffoon) [inline exam-processor/src/main/resources/application.sql.yml]: If you want to discuss data modeling and schema because of this, let me know. We'll need a big whiteboard.

File: exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcAssessmentRepository.java [hunk 9]
```diff
@@ -39,32 +38,43 @@
...
+        final List<Assessment.Builder> found = jdbcTemplate.query(sqlFindOneByNaturalId,
+                new MapSqlParameterSource("natural_id", naturalId),
+                (rs, rowNum) -> Assessment.builder()
+                        .id(rs.getInt("id"))
+                        .naturalId(rs.getString("natural_id"))
+                        .gradeCode(rs.getString("grade_code"))
+                        .typeId(rs.getInt("type_id"))
+                        .subjectId(rs.getInt("subject_id"))
+                        .subjectCode(rs.getString("subject_code")));
+        if (found.size() == 1) {
```
Comment (esotrope) [inline exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcAssessmentRepository.java]: i suppose this is nicer than the exception catching we do in reporting

File: exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcAssessmentRepository.java [hunk 10]
```diff
@@ -39,32 +38,43 @@
...
-        return jdbcTemplate.query(sqlfindSubjectClaimScores,
-                new MapSqlParameterSource()
-                        .addValue("subject_id", assessment.getSubjectId())
-                        .addValue("asmt_type_id", assessment.getTypeId()),
-                new ClaimRowMapper());
+    private List<AssessmentScore> findSubjectScores(final int asmtId) {
+        return jdbcTemplate.query(sqlfindSubjectScores,
+                new MapSqlParameterSource("asmt_id", asmtId),
+                (rs, rowNum) -> AssessmentScore.builder()
+                        .scoreType(ScoreType.valueOf(rs.getInt("score_type_id")))
```
Comment (esotrope) [inline exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcAssessmentRepository.java]: what's an `rs`?

File: migrate-olap/src/main/resources/application.warehouse.to.staging.sql.yml [hunk 11]
```diff
@@ -326,17 +326,20 @@ sql:
         # ------------ subject asmt type  ---------------------------------------------------------------------------------------
         subject_asmt_type:
           sql:
+            # TODO - deal with subject_asmt_scoring and get rid of joins here
```
Comment (malaffoon) [inline migrate-olap/src/main/resources/application.warehouse.to.staging.sql.yml]: example of migrate hack where the alt-score data is ignored while overall and claim data is transformed back to the old layout.

File: exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcAssessmentRepository.java [hunk 12]
```diff
@@ -39,32 +38,43 @@
...
+                        .subjectScoreId(getNullable(rs, rs.getInt("subject_score_id")))
+                        .code(rs.getString("code"))
+                        .minScore(getNullable(rs, rs.getDouble("min_score")))
+                        .maxScore(getNullable(rs, rs.getDouble("max_score")))
+                        .build());
     }
 
-    private Assessment.Builder appendItemData(final Assessment.Builder builder, final Assessment assessment) {
+    private Assessment.Builder appendItemData(final Assessment.Builder builder, final int asmtId) {
```
Comment (esotrope) [inline exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcAssessmentRepository.java]: is asmt short for ass mint?

File: migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml [hunk 13]
```diff
@@ -1318,12 +1318,12 @@ sql:
...
-                claim3,
-                claim4,
-                claim5,
-                claim6,
+                scs1.code,
+                scs2.code,
+                scs3.code,
+                scs4.code,
+                scs5.code,
+                scs6.code,
```
Comment (malaffoon) [inline migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml]: This is an unrelated "fix": the subselect joins are completely unnecessary.

File: exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcExamRepository.java [hunk 14]
```diff
@@ -432,49 +432,49 @@ private boolean upsertExamClaims(final List<ExamClaim> submitted, final long exa
...
+                .addValue("scale_score_std_err", examScore.getScaleScoreStdErr())
+                .addValue("theta_score", examScore.getThetaScore())
+                .addValue("theta_score_std_err", examScore.getThetaScoreStdErr())
+                .addValue("performance_level", examScore.getPerformanceLevel())
                 .getValues();
     }
 
-    private static boolean equivalent(final ExamClaim ec1, final ExamClaim ec2) {
-        return Objects.equals(ec1.getClaimId(), ec2.getClaimId())
+    private static boolean equivalent(final ExamScore ec1, final ExamScore ec2) {
```
Comment (esotrope) [inline exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcExamRepository.java]: how does equivalent differ from equals? is it looser? maybe it is doc'd i just haven't seen it yet

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/ClaimRepository.java [hunk 15]
```diff
@@ -3,7 +3,7 @@
 import java.util.Set;
 
 /**
- * Repository for dealing with the claims
+ * Repository for dealing with organizational claims.
```
Comment (malaffoon) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/ClaimRepository.java]: Just making this more obvious. Organizational claims are completely unchanged by this feature.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/SubjectAssessmentTypeRepository.java [hunk 16]
```diff
@@ -13,10 +13,10 @@
     /**
      * Return the existing subject assessment types for the given subject.
      *
-     * @param subjectCode A subject code
+     * @param subjectId A subject id
      * @return the subject assessment types or empty collection if none exist
      */
-    Collection<WarehouseSubjectAssessmentType> findBySubject(String subjectCode);
+    List<WarehouseSubjectAssessmentType> findBySubjectId(int subjectId);
```
Comment (malaffoon) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/SubjectAssessmentTypeRepository.java]: Order is important.

File: exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultStudentExamProcessor.java [hunk 17]
```diff
@@ -239,22 +238,23 @@ private Integer toValidCompleteness(final String value) {
...
     };
 
-    private Double toScaleScore(final String value) {
-        if (!hasText(value)) return null;
-
-        final Double score = Double.parseDouble(value);
-        if (score >= scoreConfiguration.getScaleScore().getMin() && score <= scoreConfiguration.getScaleScore().getMax()) return score;
+    private static Function<String, Double> scoreChecker(final AssessmentScore assessmentScore) {
+        return value -> {
+            if (!hasText(value)) return null;
```
Comment (esotrope) [inline exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultStudentExamProcessor.java]: is null being used as a signal of invalidity?

File: build.gradle (hunk 1 above)
Comment (malaffoon) [reply build.gradle]: No, it pulls my local copies. I'll pin this before committing.

File: exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcAssessmentRepository.java (hunk 10 above)
Comment (malaffoon) [reply exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcAssessmentRepository.java]: Record set. It is actually the named parameters of the function interface it is implementing. In other words, IntelliJ did that for me

File: exam-processor/src/main/resources/application.sql.yml (hunk 8 above)
Comment (esotrope) [reply exam-processor/src/main/resources/application.sql.yml]: i'll take your word it works - or more accurately the presence of tests

File: exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcExamRepository.java (hunk 14 above)
Comment (malaffoon) [reply exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcExamRepository.java]: Hmm, i was just following the pattern established in the file. But, yeah, i could probably add a comment or two.

File: migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml (hunk 13 above)
Comment (esotrope) [reply migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml]: is this because we denormalized at some point and forgot to optimize this?

File: exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultStudentExamProcessor.java (hunk 17 above)
Comment (malaffoon) [reply exam-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultStudentExamProcessor.java]: Nope: a score is allowed to be null so this is letting a blank/empty value map to no value.

File: exam-processor/src/main/resources/application.sql.yml (hunk 8 above)
Comment (malaffoon) [reply exam-processor/src/main/resources/application.sql.yml]: Yeah, the 96% code coverage *really* helped me get through these changes.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Assessment.java [hunk 18]
```diff
@@ -144,34 +108,22 @@ public Assessment build() {
...
-            this.label = assessmentToCopy.label;
-            this.version = assessmentToCopy.version;
-            this.minScore = assessmentToCopy.minScore;
-            this.maxScore = assessmentToCopy.maxScore;
-            this.cutPoint1 = assessmentToCopy.cutPoint1;
-            this.cutPoint2 = assessmentToCopy.cutPoint2;
-            this.cutPoint3 = assessmentToCopy.cutPoint3;
-            this.cutPoint4 = assessmentToCopy.cutPoint4;
-            this.cutPoint5 = assessmentToCopy.cutPoint5;
+        public Builder copyWithoutItems(final Assessment that) {
```
Comment (esotrope) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/Assessment.java]: so... when are we introducing graphql? x D

File: migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml (hunk 13 above)
Comment (malaffoon) [reply migrate-reporting/src/main/resources/application.staging.to.reporting.sql.yml]: Actually, your first comment is correct: we used to have to join because columns where in the staging tables but not the reporting tables.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/AssessmentScore.java [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+ */
+public class AssessmentScore {
+    private Integer subjectScoreId;
+    private Double minScore;
+    private Double maxScore;
+    private Double cutPoint1;
+    private Double cutPoint2;
+    private Double cutPoint3;
+    private Double cutPoint4;
+    private Double cutPoint5;
```
Comment (esotrope) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/AssessmentScore.java]: why not a list?

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/WarehouseSubjectAssessmentScoring.java [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+            return value;
+        }
+
+        public Builder copy(final Scoring scoring) {
+            type = scoring.getScoreType();
+            minScore = scoring.getMinScore();
+            maxScore = scoring.getMaxScore();
+            final PerformanceLevels levels = scoring.getPerformanceLevels();
+            if (levels != null) {
+                performanceLevelCount = levels.getPerformanceLevels().size();
```
Comment (esotrope) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/WarehouseSubjectAssessmentScoring.java]: is there a reason this isn't virtual and just delegates to `levels.getPerformanceLevels().size()` ?

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/AssessmentScore.java (hunk 19 above)
Comment (malaffoon) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/AssessmentScore.java]: Just sticking with the convention. There can be only 6 levels (5 cut-points) and we've used this approach elsewhere. Of course, this probably started when there were only 4 levels (3 cut-points) ...

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/WarehouseSubjectAssessmentScoring.java (hunk 20 above)
Comment (malaffoon) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/model/WarehouseSubjectAssessmentScoring.java]: The builder isn't keeping a copy of the `scoring` object. This isn't a normal `copy` method, it is taking an argument that is from a different data model. I waffled on this, thinking that mapping logic belongs outside the POJO but, again, this is following a convention that was already in place.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcSubjectAssessmentTypeRepository.java [hunk 21]
```diff
@@ -48,79 +49,106 @@ public JdbcSubjectAssessmentTypeRepository(final NamedParameterJdbcTemplate temp
...
-            return template.queryForObject(sqlFindBySubjectIdAndAsmtTypeId,
-                    new MapSqlParameterSource().addValue("subject_id", subjectId).addValue("asmt_type_id", asmtTypeId),
-                    this::mapToWarehouseSubjectAssessmentType);
-        } catch (final EmptyResultDataAccessException e) {
-            return null;
-        }
+        // load them all for the subject, and pull out the matching asmt type
+        return findBySubjectId(subjectId).stream()
+                .filter(t -> t.getAssessmentType().id() == asmtTypeId)
+                .findFirst().orElse(null);
```
Comment (esotrope) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcSubjectAssessmentTypeRepository.java]: feels like we should just have a tailored query but ok

Comment (malaffoon) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/repository/impl/JdbcSubjectAssessmentTypeRepository.java]: Yeah. Being a bit lazy. I'll give myself a TODO to clean this up.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultAssessmentParser.java [hunk 22]
```diff
@@ -258,18 +281,68 @@ private Assessment safeBuild(final DataElementErrorCollector elementErrorCollect
...
+         *
+         * @param record record
+         * @return list of headers
+         */
+        private List<String> discoverOverallScoreHeaders(final CSVRecord record) {
+            if (hasValue(record, "ScaledLow1") || hasValue(record, "OverallLow1")) {
+                return collectScoreHeaders(record, hasValue(record, "ScaledLow1") ? "Scaled" : "Overall");
+            } else {
+                return null;
+            }
```
Comment (esotrope) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultAssessmentParser.java]: nit: don't need else

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultPercentileParser.java [hunk 23]
```diff
@@ -138,9 +133,9 @@ void startNew(final CSVRecord record) {
 
             // get min/max score for this set of percentile ranks
             // allowing for vetting of rank scores as they come in
-            minScore = parserHelper.validate(contextName(HeaderMinScore), record.get(HeaderMinScore), toScaleScoreMin);
-            maxScore = parserHelper.validate(contextName(HeaderMaxScore), record.get(HeaderMaxScore), toScaleScoreMax);
-            if (minScore >= maxScore) {
+            minScore = parserHelper.validate(contextName(HeaderMinScore), record.get(HeaderMinScore), toDouble);
+            maxScore = parserHelper.validate(contextName(HeaderMaxScore), record.get(HeaderMaxScore), toDouble);
+            if (minScore != null && maxScore != null && minScore >= maxScore) {
```
Comment (esotrope) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultPercentileParser.java]: can they just have min or just have max?

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultSubjectService.java [hunk 24]
```diff
@@ -25,8 +52,152 @@ public void upsert(final Subject subject, final long importId) {
...
     }
+
+    private void createSubjectAssessmentTypes(final int subjectId, final Subject subject) {
+        for (final WarehouseSubjectAssessmentType type : subject.getAssessmentTypes().stream()
+                .map(t -> WarehouseSubjectAssessmentType.builder().copy(t).build()).collect(Collectors.toList())) {
+            subjectAssessmentTypeRepository.create(subjectId, type);
+        }
+    }
+
+    void updateSubjectAssessmentTypes(final int subjectId, final Subject subject) {
```
Comment (esotrope) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultSubjectService.java]: yikes. do we have coverage for all these new logic blocks?

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultSubjectTranslationService.java [hunk 25]
```diff
@@ -106,6 +117,8 @@ private void appendPerformanceLevels(final MessageContext context, final Perform
     }
 
     private void appendDepthsOfKnowledge(final MessageContext context, final List<DepthOfKnowledge> depths) {
+        if (depths == null || depths.isEmpty()) return;
```
Comment (esotrope) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultSubjectTranslationService.java]: i feel like guava has an isNullOrEmpty() maybe it is only for strings though

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/validator/AltScoreValidator.java [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+            if (!codes.add(code)) {
+                context.failure("Duplicate code found: " + code);
+                continue;
+            }
+
+            ++scorableCount;
+
+            if (score.getDisplayOrder() == null) {
+                context.failure("AltScores must have a displayOrder defined");
+            } else if (!displayOrders.add(score.getDisplayOrder())) {
```
Comment (esotrope) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/validator/AltScoreValidator.java]: not a fan of having logic in if blocks. in while-blocks it seems to be conventional but i can't say i'm a fan of that either

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/validator/PerformanceLevelValidator.java [hunk 27]
```diff
@@ -25,9 +24,18 @@
...
-            validatePerformanceLevels(context, assessmentType.getPerformanceLevels());
+            validatePerformanceLevels(context, assessmentType.getOverallScoring().getPerformanceLevels());
             if ("iab".equals(assessmentType.getCode())) {
-                validateIabPerformanceLevels(context, assessmentType.getPerformanceLevels());
+                validateIabPerformanceLevels(context, assessmentType.getOverallScoring().getPerformanceLevels());
+            }
+
+            if (assessmentType.getAltScoring() != null) {
+                context.push("AltScoring");
+                if ("iab".equals(assessmentType.getCode())) {
```
Comment (esotrope) [inline package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/validator/PerformanceLevelValidator.java]: i think common has AssessmentType.IAB.code() or something

File: package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/model/WarehouseSubjectAssessmentScoringTest.java [hunk 28]
```diff
@@ -0,0 +1,39 @@
...
+                .performanceLevels(PerformanceLevels.builder()
+                        .standardCutoff(1)
+                        .performanceLevels(ImmutableList.of(
+                                PerformanceLevel.builder().level(1).build(),
+                                PerformanceLevel.builder().level(2).build()))
+                        .build())
+                .build();
+        final WarehouseSubjectAssessmentScoring scoring = WarehouseSubjectAssessmentScoring.builder().copy(src).build();
+
+        assertThat(scoring.getType()).isEqualTo(ScoreType.CLAIM);
```
Comment (esotrope) [inline package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/model/WarehouseSubjectAssessmentScoringTest.java]: why not use the recursive field-by-field?

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultAssessmentParser.java (hunk 22 above)
Comment (malaffoon) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultAssessmentParser.java]: will remove

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultPercentileParser.java (hunk 23 above)
Comment (malaffoon) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultPercentileParser.java]: Sure, though not likely. If either is null, the error collector will have already had an error added so the overall import will fail. This is in keeping with the approach of collecting as many errors as possible before rejecting the import.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultSubjectService.java (hunk 24 above)
Comment (malaffoon) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultSubjectService.java]: Yup, all covered.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultSubjectTranslationService.java (hunk 25 above)
Comment (malaffoon) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/DefaultSubjectTranslationService.java]: Yeah, i've been meaning to sweep through and make consistent use of guava (or commons). I had basically the same comment when this code originally came through (FTR, this isn't really new code, just massaged existing code).

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/validator/AltScoreValidator.java (hunk 26 above)
Comment (malaffoon) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/validator/AltScoreValidator.java]: Hmm, i think it looks better than the extra indentation. We can thumb wrestle over it.

File: package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/validator/PerformanceLevelValidator.java (hunk 27 above)
Comment (malaffoon) [reply package-processor/src/main/java/org/opentestsystem/rdw/ingest/processor/service/impl/validator/PerformanceLevelValidator.java]: Good point, will change.

File: package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/model/WarehouseSubjectAssessmentScoringTest.java (hunk 28 above)
Comment (malaffoon) [reply package-processor/src/test/java/org/opentestsystem/rdw/ingest/processor/model/WarehouseSubjectAssessmentScoringTest.java]: They are different object models. I guess i could manually construct an `expected` scoring but this seemed easier.

**Comment by malaffoon** [comment]
> Looks greate and I wasn't able to detect anything "wrong" so it's good for me i think.

Thanks for grinding through it!
